### PR TITLE
 Ensure generated task and step names are RFC 1035 compliant DNS labels

### PR DESCRIPTION
**Pull request by dwnusbaum**
Still needs some unit tests for the new function at the least, but I wanted to go ahead and file what I had. I also have a few questions on how we want to handle some edge cases.

### CODE REVIEW

File: pkg/buildpipeline/pipeline.go [hunk 1]
```diff
@@ -155,6 +155,61 @@ func (s *Stage) TaskName() string {
...
+// cf. https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+// Suffix is assumed to be alphanumeric and non-empty.
+func createTaskName(pipelineIdentifier string, stageName string, buildIdentifier string, suffix string) string {
+	const MAX_LABEL_LENGTH = 63
+	MAX_BODY_LENGTH := MAX_LABEL_LENGTH - len(suffix) - 1 // Add an extra hyphen before the suffix
+
+	// Having `stage` and `build` literally in the output means the body is
+	// guaranteed to be non-empty. Would it be better to fail if the output is
+	// `stage-build-$suffix` because everything else is invalid?
+	body := fmt.Sprintf("%s-stage-%s-build-%s", pipelineIdentifier, stageName, buildIdentifier)
```
Comment (dwnusbaum) [inline pkg/buildpipeline/pipeline.go]: For example, see my test case using `. --- .` as a stage name of a case where failing might make more sense.

Comment (dwnusbaum) [reply pkg/buildpipeline/pipeline.go]: Depending on how long our suffixes end up being we could use something like `%s-s-%s-b-%s` to reclaim a few characters.

Also, it feels a little clearer to me to use a pipeline-build-stage order, so that any sorted output would automatically put tasks related to a single build together vs putting stages together across multiple builds.

File: pkg/buildpipeline/pipeline.go [hunk 2]
```diff
@@ -519,6 +574,7 @@ func stageToTask(s Stage, pipelineIdentifier string, buildIdentifier string, nam
 					stepImage = step.Agent.Image
 				}
 				t.Spec.Steps = append(t.Spec.Steps, corev1.Container{
+					// TODO: Step names don't appear to be validated in Build Pipeline _currently_, but should they be?
```
Comment (dwnusbaum) [inline pkg/buildpipeline/pipeline.go]: Nvm, they need to be RFC 1035 labels as well: https://github.com/knative/build-pipeline/blob/2ce21d9bba67fc3e11122d33bd5b4e1b1a044b9c/pkg/apis/pipeline/v1alpha1/task_validation.go#L73.

File: pkg/buildpipeline/pipeline.go (hunk 1 above)
Comment (abayer) [reply pkg/buildpipeline/pipeline.go]: Yeah, I'd say we should require at least *some* alphanumeric characters.

Comment (abayer) [reply pkg/buildpipeline/pipeline.go]: and I'm utterly indifferent to the order - I was just throwing stuff around. =)

File: pkg/buildpipeline/pipeline.go (hunk 2 above)
Comment (abayer) [reply pkg/buildpipeline/pipeline.go]: Makes sense, since each step results in a container.

File: pkg/buildpipeline/pipeline.go [hunk 3]
```diff
@@ -155,6 +155,61 @@ func (s *Stage) TaskName() string {
...
+	MAX_BODY_LENGTH := MAX_LABEL_LENGTH - len(suffix) - 1 // Add an extra hyphen before the suffix
+
+	// Having `stage` and `build` literally in the output means the body is
+	// guaranteed to be non-empty. Would it be better to fail if the output is
+	// `stage-build-$suffix` because everything else is invalid?
+	body := fmt.Sprintf("%s-stage-%s-build-%s", pipelineIdentifier, stageName, buildIdentifier)
+
+	var sb strings.Builder
+	bufferedHyphen := false // Used to make sure we don't output consecutive hyphens.
+	for _, codepoint := range body {
```
Comment (abayer) [inline pkg/buildpipeline/pipeline.go:171]: I gotta wonder whether there might be something like this in some library we already pull in...but hey.

Comment (dwnusbaum) [reply pkg/buildpipeline/pipeline.go:171]: Yeah very possible. There is https://github.com/kubernetes/kubernetes/blob/0b8566f3887a2d13baba623d88aeb64f6e637f46/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L162 to just say "is this valid?", but I didn't see any mangling functions.

File: pkg/buildpipeline/pipeline.go (hunk 2 above)
Comment (dwnusbaum) [reply pkg/buildpipeline/pipeline.go]: Done in 0891c988961

File: pkg/buildpipeline/pipeline.go (hunk 1 above)
Comment (dwnusbaum) [reply pkg/buildpipeline/pipeline.go]: Ok, I went ahead and added validation in 2323fb5 and switched up the order in aadcfef.

**Review by abayer — APPROVED** [review]

File: pkg/buildpipeline/pipeline.go (hunk 3 above)
Comment (dwnusbaum) [reply pkg/buildpipeline/pipeline.go:171]: Turns out JX already a function that does most of what we want, but it doesn't make capital letters lowercase or enforce a maximum length: https://github.com/jenkins-x/jx/blob/43f6dae88b40a139572758149d92df6d366796ab/pkg/kube/names.go#L18.
